File: 26-sum/prog/td-game/src/game/wave.cpp
#include "wave.h"

#include "factory.h"
#include "game.h"
#include "observer.h"

int WaveManager::current_wave() const { return current_wave_ + 1; }
bool WaveManager::all_waves_done() const { return all_done_; }

void WaveManager::start(WaveSpec const& spec) {
  spec_ = spec;
  current_wave_ = -1;
  wave_time_ = 0.0f;
  all_done_ = false;
  if (!spec_.waves || spec_.waves->empty()) {
    in_gap_ = false;
    all_done_ = true;
  } else {
    enter_gap(0);
  }
}

void WaveManager::enter_gap(int index) {
  in_gap_ = true;
  gap_timer_ = CountdownTimer((*spec_.gaps)[static_cast<std::size_t>(index)], true);
}

void WaveManager::start_wave(Game& game, int index) {
  current_wave_ = index;
  in_gap_ = false;
  wave_time_ = 0.0f;

  auto const& spawns = (*spec_.waves)[static_cast<std::size_t>(index)].spawns;
  spawned_.assign(spawns.size(), false);

  int wave_count = static_cast<int>(spec_.waves->size());
  if (index >= 0 && index < wave_count) {
    bool has_boss = false;
    bool is_last = (index == wave_count - 1);

    for (auto const& s : spawns) {
      if (s.type == "boss") {
        has_boss = true;
        break;
      }
    }

    for (auto* o : game.observers()) o->on_wave_started(index + 1, has_boss, is_last);
  }
}

void WaveManager::update(Game& game, float dt) {
  if (all_done_ || !spec_.waves) return;

  // Pre-wave gap phase: wait gaps[current_wave_] before the wave starts.
  if (in_gap_) {
    gap_timer_.update(dt);
    if (gap_timer_.is_finished()) {
      start_wave(game, current_wave_ + 1);
    }
    return;
  }

  if (current_wave_ >= static_cast<int>(spec_.waves->size())) {
    all_done_ = true;
    return;
  }

  wave_time_ += dt;
  Wave const& wave = (*spec_.waves)[static_cast<std::size_t>(current_wave_)];
  bool all_spawned = true;

  for (std::size_t i = 0; i < wave.spawns.size(); ++i) {
    if (!spawned_[i] && wave_time_ >= wave.spawns[i].time) {
      spawned_[i] = true;
      int route = wave.spawns[i].route;  // validated at load time

      auto enemy = make_enemy(
        wave.spawns[i].type,
        spec_.enemy_stats->at(wave.spawns[i].type),
        spec_.routes[static_cast<std::size_t>(route)]
      );
      game.spawn_enemy(std::move(enemy));
    }

    if (!spawned_[i]) all_spawned = false;
  }

  // Once all spawns of this wave are issued, enter the gap before the next wave
  // (or finish if this was the last wave).
  if (all_spawned) {
    int next = current_wave_ + 1;
    if (next >= static_cast<int>(spec_.waves->size())) {
      all_done_ = true;
    } else {
      enter_gap(next);
    }
  }
}

File: 26-sum/prog/td-game/src/rust/src/wave.rs
use crate::{
    config::{EnemyConfigTable, make_enemy},
    enemy::Enemy,
    path::Path,
    timer::CountdownTimer,
};

/// A wave is a sequence of timed spawns, with prescribed path for each spawn.
pub type Wave = Vec<(String, f32, usize)>;

/// The wave schedule and the per-wave pre-delays.
/// `gaps[0]` elapses from game start to wave 0; `gaps[i]` (i>0) elapses between
/// the previous wave's last spawn and wave i.
#[derive(Debug, Clone)]
pub struct WaveSpec {
    pub waves: Vec<Wave>,
    pub gaps: Vec<f32>,
    pub current_wave: usize,
    pub all_done: bool,
    /// Seconds since the current wave started spawning.
    wave_time: f32,
    /// Which spawns of the current wave have fired.
    spawned: Vec<bool>,
    gap_timer: CountdownTimer,
    in_gap: bool,
}

impl WaveSpec {
    pub fn new(waves: Vec<Wave>, gaps: Vec<f32>) -> Self {
        assert_eq!(waves.len(), gaps.len());
        Self {
            all_done: waves.is_empty(), // put this first to pass the borrow checker
            waves,
            gaps,
            current_wave: 0,
            wave_time: 0.0,
            spawned: vec![],
            gap_timer: CountdownTimer::new(0.0, false),
            in_gap: false,
        }
    }

    pub fn current_wave_display(&self) -> i32 {
        if self.in_gap {
            self.current_wave as i32
        } else {
            (self.current_wave as i32 + 1).min(self.waves.len() as i32)
        }
    }

    pub fn start(&mut self) {
        if !self.waves.is_empty() {
            // Enter gap 0.
            self.in_gap = true;
            self.gap_timer = CountdownTimer::new(self.gaps[0], true);
        }
    }

    pub fn update(
        &mut self,
        dt: f32,
        enemy_configs: &EnemyConfigTable,
        paths: &[Path],
        mut spawn_enemy: impl FnMut(Enemy),
        mut emit_wave_started: impl FnMut(i32, bool, bool), // index, has_boss, is_last
    ) {
        if self.waves.is_empty() || self.all_done {
            return;
        }

        if self.in_gap {
            if self.gap_timer.update(dt) {
                self.in_gap = false;
                self.wave_time = 0.0;
                self.spawned.clear();
                self.spawned
                    .resize(self.waves[self.current_wave].len(), false);

                emit_wave_started(
                    self.current_wave as i32 + 1,
                    self.waves[self.current_wave]
                        .iter()
                        .any(|(kind, _, _)| kind == "boss"),
                    self.current_wave == self.waves.len() - 1,
                )
            }
        } else {
            self.wave_time += dt;

            let all_spawned = self.waves[self.current_wave]
                .iter()
                .zip(self.spawned.iter_mut())
                .fold(true, |acc, ((kind, time, path_index), spawned)| {
                    if !*spawned && self.wave_time >= *time {
                        if let Some(enemy) = make_enemy(kind, enemy_configs, *path_index, paths) {
                            spawn_enemy(enemy);
                        }
                        *spawned = true;
                    }
                    acc && *spawned
                });

            if all_spawned {
                self.current_wave += 1;
                if self.current_wave >= self.waves.len() {
                    self.all_done = true;
                } else {
                    self.in_gap = true;
                    self.gap_timer = CountdownTimer::new(self.gaps[self.current_wave], true);
                }
            }
        }
    }
}
